In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

In [2]:
%matplotlib inline

### number 5: Using the pandas read_csv() method, read the GDP dataset into your notebook as a DataFrame called gdp_df. Take a look at the first 6 rows.

In [3]:
gdp_df = pd.read_csv("../data/gdp_percapita.csv", nrows=6869)

##### tip from Maggie: footnotes begin in row 6869 in this file, add nrows into read_csv code

In [4]:
print(gdp_df.head(6))

  Country or Area  Year        Value  Value Footnotes
0     Afghanistan  2019  2065.036398              NaN
1     Afghanistan  2018  2033.779002              NaN
2     Afghanistan  2017  2058.383832              NaN
3     Afghanistan  2016  2057.062164              NaN
4     Afghanistan  2015  2068.265904              NaN
5     Afghanistan  2014  2102.385234              NaN


### number 6: Repeat for the internet use dataset. Call this DataFrame internet_df. Take a look at the first six rows.

In [5]:
internet_df = pd.read_csv("../data/internet_use.csv", nrows=4496)

##### from watching classmates in setup demo, know to use nrows to avoid footnote error

In [6]:
print(internet_df.head(6))

  Country or Area  Year     Value  Value Footnotes
0     Afghanistan  2014  6.390000              NaN
1     Afghanistan  2013  5.900000              NaN
2     Afghanistan  2012  5.454545              NaN
3     Afghanistan  2011  5.000000              NaN
4     Afghanistan  2010  4.000000              NaN
5     Afghanistan  2009  3.550000              NaN


### number 7: Look at the shape of each DataFrame - how many rows? How many columns?

In [7]:
gdp_df.shape

(6869, 4)

In [8]:
internet_df.shape

(4496, 4)

the GDP table has 6869 rows and 4 columns; 
the internet table has 4496 rows and 4 columns

### number 8: Take a look at the datatypes for the columns in each DataFrame.

In [9]:
print(type(gdp_df))

<class 'pandas.core.frame.DataFrame'>


In [10]:
print(gdp_df.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6869 entries, 0 to 6868
Data columns (total 4 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   Country or Area  6869 non-null   object 
 1   Year             6869 non-null   object 
 2   Value            6868 non-null   float64
 3   Value Footnotes  1 non-null      float64
dtypes: float64(2), object(2)
memory usage: 214.8+ KB
None


In [11]:
print(internet_df.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4496 entries, 0 to 4495
Data columns (total 4 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   Country or Area  4496 non-null   object 
 1   Year             4496 non-null   object 
 2   Value            4495 non-null   float64
 3   Value Footnotes  948 non-null    float64
dtypes: float64(2), object(2)
memory usage: 140.6+ KB
None


Both the internet and GDP dataframes have four columns -- the first two of which are object types, and the latter two are float64 types

### number 9: Take a look at the last 10 rows of each DataFrame in turn.

In [12]:
print(gdp_df.tail(10))

     Country or Area      Year        Value  Value Footnotes
6859        Zimbabwe      1998  3653.920016              NaN
6860        Zimbabwe      1997  3580.048793              NaN
6861        Zimbabwe      1996  3520.430146              NaN
6862        Zimbabwe      1995  3226.413930              NaN
6863        Zimbabwe      1994  3263.934978              NaN
6864        Zimbabwe      1993  3033.504852              NaN
6865        Zimbabwe      1992  3054.889178              NaN
6866        Zimbabwe      1991  3426.598094              NaN
6867        Zimbabwe      1990  3324.348171              NaN
6868   footnoteSeqID  Footnote          NaN              NaN


In [13]:
print(internet_df.tail(10))

     Country or Area      Year     Value  Value Footnotes
4486        Zimbabwe      2001  0.799846              NaN
4487        Zimbabwe      2000  0.401434              NaN
4488        Zimbabwe      1999  0.161676              NaN
4489        Zimbabwe      1998  0.081648              NaN
4490        Zimbabwe      1997  0.033080              NaN
4491        Zimbabwe      1996  0.016790              NaN
4492        Zimbabwe      1995  0.007684              NaN
4493        Zimbabwe      1994  0.001739              NaN
4494        Zimbabwe      1990  0.000000              NaN
4495   footnoteSeqID  Footnote       NaN              NaN


### number 10: Drop the value footnotes column from both DataFrames. Check that this worked as expected.

In [14]:
gdp_df_dropped = gdp_df.drop(columns = 'Value Footnotes')

In [15]:
print(gdp_df_dropped.head(5))

  Country or Area  Year        Value
0     Afghanistan  2019  2065.036398
1     Afghanistan  2018  2033.779002
2     Afghanistan  2017  2058.383832
3     Afghanistan  2016  2057.062164
4     Afghanistan  2015  2068.265904


In [16]:
print(gdp_df.head(5))

  Country or Area  Year        Value  Value Footnotes
0     Afghanistan  2019  2065.036398              NaN
1     Afghanistan  2018  2033.779002              NaN
2     Afghanistan  2017  2058.383832              NaN
3     Afghanistan  2016  2057.062164              NaN
4     Afghanistan  2015  2068.265904              NaN


##### okay, so... it worked to make a new dataframe that's defined as the dropped dataframe... I'm going to pause and go read more about dropping columns to make sure I'm understanding A) what's happening here and B) what the question wants me to achieve.

##### adding a line of markdown so gitbash will let me add this unchanged file because I worked on datacamp all night :)

In [17]:
internet_df_dropped = internet_df.drop(columns = 'Value Footnotes')

In [18]:
print(internet_df_dropped.head(5))

  Country or Area  Year     Value
0     Afghanistan  2014  6.390000
1     Afghanistan  2013  5.900000
2     Afghanistan  2012  5.454545
3     Afghanistan  2011  5.000000
4     Afghanistan  2010  4.000000


In [19]:
print(internet_df.head(5))

  Country or Area  Year     Value  Value Footnotes
0     Afghanistan  2014  6.390000              NaN
1     Afghanistan  2013  5.900000              NaN
2     Afghanistan  2012  5.454545              NaN
3     Afghanistan  2011  5.000000              NaN
4     Afghanistan  2010  4.000000              NaN


##### out of an abundance of caution, I'm going to keep all four dataframes -- gdp w/ and w/o drop, internet w/ and w/o drop.

### number 11: Change the columns for the GDP Per Capita DataFrame to ‘Country’, ‘Year’, and ‘GDP_Per_Capita’.

##### addendum to decision to keep all 4 dataframes -- going to rename the _dropped ones to the brief names and rename the undropped ones to _original

In [20]:
gdp_df_original = gdp_df

In [21]:
internet_df_original = internet_df

In [22]:
gdp_df = gdp_df_dropped

In [23]:
internet_df = internet_df_dropped

##### now to rename the columns

In [24]:
gdp_df.rename({'Country or Area' : 'Country', 'Year' : 'Year', 'Value' : 'GDP_Per_Capita'}, axis = 1, inplace = True)

In [25]:
print(gdp_df.head(5))

       Country  Year  GDP_Per_Capita
0  Afghanistan  2019     2065.036398
1  Afghanistan  2018     2033.779002
2  Afghanistan  2017     2058.383832
3  Afghanistan  2016     2057.062164
4  Afghanistan  2015     2068.265904


### number 12: Change the columns for the Internet Users DataFrame to ‘Country’, ‘Year’, and ‘Internet_Users_Pct’.

In [26]:
internet_df.rename({'Country or Area' : 'Country', 'Year' : 'Year', 'Value' : 'Internet_Users_Pct'}, axis = 1, inplace = True)

In [27]:
print(internet_df.head(5))

       Country  Year  Internet_Users_Pct
0  Afghanistan  2014            6.390000
1  Afghanistan  2013            5.900000
2  Afghanistan  2012            5.454545
3  Afghanistan  2011            5.000000
4  Afghanistan  2010            4.000000


### number 13: Merge the two DataFrames to one. Merge all rows from each of the two DataFrames. Call the new DataFrame gdp_and_internet_use.

In [28]:
gdp_and_internet_use = pd.merge(gdp_df, internet_df, on=['Country', 'Year'])

note: this is incorrect, it's redone later, I'm keeping it for posterity

### number 14: Look at the first five rows of your new DataFrame to confirm it merged correctly.

In [29]:
print(gdp_and_internet_use.head(5))

       Country  Year  GDP_Per_Capita  Internet_Users_Pct
0  Afghanistan  2014     2102.385234            6.390000
1  Afghanistan  2013     2116.465454            5.900000
2  Afghanistan  2012     2075.491747            5.454545
3  Afghanistan  2011     1904.559799            5.000000
4  Afghanistan  2010     1957.029338            4.000000


spoiler: I didn't, it is incorrect, it's redone later, I'm keeping it for posterity

### number 15: Look at the last five rows to make sure the data is clean and as expected.

In [30]:
print(gdp_and_internet_use.tail(5))

            Country      Year  GDP_Per_Capita  Internet_Users_Pct
3454       Zimbabwe      1996     3520.430146            0.016790
3455       Zimbabwe      1995     3226.413930            0.007684
3456       Zimbabwe      1994     3263.934978            0.001739
3457       Zimbabwe      1990     3324.348171            0.000000
3458  footnoteSeqID  Footnote             NaN                 NaN


spoiler: I didn't it is incorrect, it's redone later, I'm keeping it for posterity

##### based on the next question being about subsetting to only certain years, I'm not entirely concerned about the final row being whacky. I'm a bit torn about the join type -- left, right, inner, outer, etc -- because in theory, since the GDP dataframe has 6,869 rows and the internet dataframe has 4,496, the 3,457 rows in the merged dataframe is both unexpected and perhaps concerning. I imagine it's because there are years of missing data between the two DFs (some years where GDP data is recorded but internet is not, even in countries that are in both DFs, etc), but... should I have done an outer join?

##### hold the phone, the text of question 13 literally says "Merge all rows" I'm a silly goose. OKAY OUTER JOIN IT IS -- time to re-do #13 onward

### question 13 TAKE TWO: Merge the two DataFrames to one. Merge all rows from each of the two DataFrames. Call the new DataFrame gdp_and_internet_use.

In [31]:
gdp_and_internet_use = pd.merge(gdp_df, internet_df, on=['Country', 'Year'], how='outer')

### question 14 TAKE TWO: Look at the first five rows of your new DataFrame to confirm it merged correctly.

In [32]:
print(gdp_and_internet_use.head(5))

       Country  Year  GDP_Per_Capita  Internet_Users_Pct
0  Afghanistan  2019     2065.036398                 NaN
1  Afghanistan  2018     2033.779002                 NaN
2  Afghanistan  2017     2058.383832                 NaN
3  Afghanistan  2016     2057.062164                 NaN
4  Afghanistan  2015     2068.265904                 NaN


### question 15 TAKE TWO: Look at the last five rows to make sure the data is clean and as expected.


In [33]:
print(gdp_and_internet_use.tail(5))

     Country  Year  GDP_Per_Capita  Internet_Users_Pct
7901   Yemen  1999             NaN            0.056629
7902   Yemen  1998             NaN            0.023323
7903   Yemen  1997             NaN            0.015025
7904   Yemen  1996             NaN            0.000621
7905   Yemen  1990             NaN            0.000000


##### YAY LOOK AT IT WOW!!! The last row is clean, there's more rows than the So many null values though... I'm going to keep them and assume it's fine. I see .fillna options and whatnot online, but I'll cross that bridge when I get to it. But, also... where is Zimbabwe?

In [34]:
print(gdp_and_internet_use.tail(50))

                    Country  Year  GDP_Per_Capita  Internet_Users_Pct
7856               Viet Nam  2002             NaN            1.854999
7857               Viet Nam  2001             NaN            1.265651
7858               Viet Nam  2000             NaN            0.254248
7859               Viet Nam  1999             NaN            0.128927
7860               Viet Nam  1998             NaN            0.013079
7861               Viet Nam  1997             NaN            0.003982
7862               Viet Nam  1996             NaN            0.000135
7863               Viet Nam  1990             NaN            0.000000
7864  Virgin Islands (U.S.)  2014             NaN           50.070000
7865  Virgin Islands (U.S.)  2013             NaN           45.300000
7866  Virgin Islands (U.S.)  2012             NaN           40.547900
7867  Virgin Islands (U.S.)  2011             NaN           35.600000
7868  Virgin Islands (U.S.)  2010             NaN           31.220000
7869  Virgin Islands

##### So looking at the bottom 5 rows of both dataframes, I was expecting the last row of my merged dataframe to be
##### Zimbabwe     ____      1990   ____     3324.348171   ____      0.0000
##### but... it's not... and I don't know why

In [35]:
print(gdp_df.tail(60))
print(internet_df.tail(60))

            Country      Year  GDP_Per_Capita
6809         Zambia      2018     3521.520144
6810         Zambia      2017     3485.005238
6811         Zambia      2016     3467.875156
6812         Zambia      2015     3443.555206
6813         Zambia      2014     3450.046561
6814         Zambia      2013     3399.712999
6815         Zambia      2012     3339.276135
6816         Zambia      2011     3201.289398
6817         Zambia      2010     3125.529308
6818         Zambia      2009     2917.516220
6819         Zambia      2008     2747.438718
6820         Zambia      2007     2619.721370
6821         Zambia      2006     2483.208778
6822         Zambia      2005     2362.901628
6823         Zambia      2004     2261.766537
6824         Zambia      2003     2168.332619
6825         Zambia      2002     2080.196828
6826         Zambia      2001     2042.538262
6827         Zambia      2000     1990.858609
6828         Zambia      1999     1968.215803
6829         Zambia      1998     

##### okay, so it looks like Zimbabwe and Zambia were both cut off, but Yemen looks complete -- the merged DF has Yemen data from 1990, which is the lowest in the internet usage DF (and I don't think I've seen a year lower than 1990 so it feels promising)

##### after some research and a quick conference with the instructors, I believe that the order of operations with an outer join is [merge rows with data in both DFs], then [merge rows with data in first DF but NOT the second one], THEN [merge rows with data in second DF but NOT the first one]
##### implication: the Zimbabwe ____ 1990 ____ 3324.348171 ____ 0.0000 row is probably somewhere in the middle. I'm going to try to pull out rows where Country = Zimbabwe or Zambia just to make sure they exist, then assume that I have the right merged DF...

In [38]:
merge_check = gdp_and_internet_use[gdp_and_internet_use.Country == 'Zimbabwe']

In [39]:
print(merge_check)

       Country  Year  GDP_Per_Capita  Internet_Users_Pct
6838  Zimbabwe  2019     2835.948532                 NaN
6839  Zimbabwe  2018     3130.029501                 NaN
6840  Zimbabwe  2017     3028.245976                 NaN
6841  Zimbabwe  2016     2934.733650                 NaN
6842  Zimbabwe  2015     2958.209004                 NaN
6843  Zimbabwe  2014     2955.240287           19.890000
6844  Zimbabwe  2013     2937.725624           18.500000
6845  Zimbabwe  2012     2932.081336           17.090000
6846  Zimbabwe  2011     2556.278431           15.700000
6847  Zimbabwe  2010     2273.200073           11.500000
6848  Zimbabwe  2009     1925.364381           11.360000
6849  Zimbabwe  2008     1739.242807           11.400000
6850  Zimbabwe  2007     2133.808186           10.850000
6851  Zimbabwe  2006     2233.017549            9.791842
6852  Zimbabwe  2005     2328.176129            8.015978
6853  Zimbabwe  2004     2480.859342            6.564045
6854  Zimbabwe  2003     2642.1

##### SUCCESS!!! Moving on.

### number 16: Subset the combined DataFrame to keep only the data for 2004, 2009, and 2014. Check that this happened correctly.

In [40]:
years_df = gdp_and_internet_use[gdp_and_internet_use['Year'].isin(['2004', '2009', '2014'])]

In [43]:
print(years_df)

                    Country  Year  GDP_Per_Capita  Internet_Users_Pct
5               Afghanistan  2014     2102.385234            6.390000
10              Afghanistan  2009     1758.904043            3.550000
15              Afghanistan  2004     1200.278256            0.105809
23                  Albania  2014    11623.866679           60.100000
28                  Albania  2009    10346.864199           41.200000
...                     ...   ...             ...                 ...
7869  Virgin Islands (U.S.)  2009             NaN           27.396510
7874  Virgin Islands (U.S.)  2004             NaN           27.377009
7886                  Yemen  2014             NaN           22.550000
7891                  Yemen  2009             NaN            9.960000
7896                  Yemen  2004             NaN            0.881223

[830 rows x 4 columns]


### number 17: Create three new DataFrames, one for 2004, one for 2009, and one for 2014. Give them meaningful names that aren't too long.

In [44]:
df_2004 = years_df[years_df.Year == '2004']

In [45]:
print(df_2004)

                    Country  Year  GDP_Per_Capita  Internet_Users_Pct
15              Afghanistan  2004     1200.278256            0.105809
33                  Albania  2004     7604.838002            2.420388
63                  Algeria  2004    10058.435939            4.634475
93                   Angola  2004     5593.241537            0.464815
123     Antigua and Barbuda  2004    18745.681534           24.266544
...                     ...   ...             ...                 ...
7772     T.F.Y.R. Macedonia  2004             NaN           24.440000
7830              Venezuela  2004             NaN            8.404470
7854               Viet Nam  2004             NaN            7.642409
7874  Virgin Islands (U.S.)  2004             NaN           27.377009
7896                  Yemen  2004             NaN            0.881223

[276 rows x 4 columns]


In [46]:
df_2009 = years_df[years_df.Year == '2009']

In [47]:
print(df_2009)

                    Country  Year  GDP_Per_Capita  Internet_Users_Pct
10              Afghanistan  2009     1758.904043             3.55000
28                  Albania  2009    10346.864199            41.20000
58                  Algeria  2009    10782.361398            11.23000
88                   Angola  2009     7643.493749             6.00000
118     Antigua and Barbuda  2009    20046.839017            42.00000
...                     ...   ...             ...                 ...
7767     T.F.Y.R. Macedonia  2009             NaN            51.77000
7825              Venezuela  2009             NaN            32.70000
7849               Viet Nam  2009             NaN            26.55000
7869  Virgin Islands (U.S.)  2009             NaN            27.39651
7891                  Yemen  2009             NaN             9.96000

[277 rows x 4 columns]


In [48]:
df_2014 = years_df[years_df.Year == '2014']

In [49]:
print(df_2014)

                    Country  Year  GDP_Per_Capita  Internet_Users_Pct
5               Afghanistan  2014     2102.385234                6.39
23                  Albania  2014    11623.866679               60.10
53                  Algeria  2014    11512.705405               18.09
83                   Angola  2014     8239.828837               21.26
113     Antigua and Barbuda  2014    18104.241470               64.00
...                     ...   ...             ...                 ...
7762     T.F.Y.R. Macedonia  2014             NaN               68.06
7820              Venezuela  2014             NaN               57.00
7844               Viet Nam  2014             NaN               48.31
7864  Virgin Islands (U.S.)  2014             NaN               50.07
7886                  Yemen  2014             NaN               22.55

[277 rows x 4 columns]


### number 18: Which country had the highest percentage of internet users in 2014? What was the percentage? (Try typing the first 3 letters of your DataFrame name and hitting the tab key for auto-complete options).

##### so my research is showing approximately 10,000 ways to do this, so I may need to try a few... argmax(), df.max(), define the column alone and do column.max, lots of ways... let's try this one first

In [50]:
df_2014[df_2014['Internet_Users_Pct']==df_2014['Internet_Users_Pct'].max()]

Country  Year  GDP_Per_Capita  Internet_Users_Pct
2625  Iceland  2014      49652.1495               98.16

I want to do more research and find other ways to do this to make sure this answer is correct, but I think it's exactly what I wanted... but I'll pin it for now!